File: src/entity/Publication.ts
/**
 * @main Publication
 */

/**
 * @title Publication
 */
type Publication = {
  /**
   * An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  /**
   * The publication category.
   */
  category: Category

  /**
   * The age the content is restricted to.
   */
  age_restriction: AgeRestriction

  /**
   * The publication is not (fully) implemented and thus needs to be excluded
   * from * stable releases.
   */
  is_missing_implementation?: boolean

  /**
   * The specific other publications this publication depends on data from.
   */
  prerequisites?: Prerequisite[]

  /**
   * All translations for the entry, identified by IETF language tag (BCP47).
   * @minProperties 1
   */
  translations: {
    /**
     * @patternProperties ^[a-z]{2}-[A-Z]{2}$
     */
    [localeId: string]: Translation
  }
}

/**
 * The publication category.
 */
type Category =
  | { tag: "CoreRules" }
  | { tag: "ExpansionRules" }
  | { tag: "Sourcebook" }
  | { tag: "RegionalSourcebook" }

/**
 * The age the content is restricted to.
 */
type AgeRestriction =
  | { tag: "Adult" }
  | { tag: "None" }

type Prerequisite = {
  /**
   * The publication identifier.
   * @integer
   * @minimum 1
   */
  id: number
}

type Translation = {
  /**
   * The publisher's publication identifier.
   * @minLength 1
   * @example US25001
   */
  id?: string

  /**
   * The publication's name.
   * @minLength 1
   */
  name: string

  /**
   * The publication's abbreviation.
   * @minLength 1
   */
  abbreviation: string

  /**
   * The publication's release date.
   * @format date
   */
  release_date?: string

  /**
   * This publication translation is not (fully) implemented and thus needs to
   * be excluded from stable releases.
   */
  is_missing_implementation?: boolean
}


// {
//   "$schema": "http://json-schema.org/draft-07/schema",
//   "$id": "src/entity/publication/Publication.schema.json",
//   "title": "Publication",
//   "type": "object",
//   "properties": {
//     "id": {
//       "description": "An unique, increasing integer.",
//       "type": "integer",
//       "minimum": 1
//     },
//     "category": {
//       "description": "The publication category.",
//       "oneOf": [
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "CoreRules" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         },
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "ExpansionRules" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         },
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "Sourcebook" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         },
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "RegionalSourcebook" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         }
//       ]
//     },
//     "age_restriction": {
//       "description": "The age the content is restricted to.",
//       "oneOf": [
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "Adult" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         },
//         {
//           "type": "object",
//           "properties": {
//             "tag": { "const": "None" }
//           },
//           "required": ["tag"],
//           "additionalProperties": false
//         }
//       ]
//     },
//     "is_missing_implementation": {
//       "description": "The book is not (fully) implemented and thus needs to be excluded from the stable releases.",
//       "type": "boolean"
//     },
//     "prerequisites": {
//       "description": "The specific other publications this publication depends on data from.",
//       "type": "array",
//       "items": {
//         "type": "object",
//         "properties": {
//           "id": {
//             "description": "The publication identifier.",
//             "type": "integer",
//             "minimum": 1
//           }
//         },
//         "required": ["id"],
//         "additionalProperties": false
//       },
//       "minItems": 1,
//       "uniqueItems": true
//     },
//     "translations": {
//       "type": "object",
//       "description": "All translations for the entry, identified by IETF language tag (BCP47).",
//       "patternProperties": {
//         "^[a-z]{2}-[A-Z]{2}$": {
//           "type": "object",
//           "properties": {
//             "id": {
//               "description": "The book's publisher ID.",
//               "type": "string",
//               "minLength": 1
//             },
//             "name": {
//               "description": "The name of the book.",
//               "type": "string",
//               "minLength": 1
//             },
//             "abbreviation": {
//               "description": "The abbreviation of the book's name.",
//               "type": "string",
//               "minLength": 1
//             },
//             "printings": {
//               "$comment": "Experimental API",
//               "description": "The number of printings of this publication.",
//               "type": "integer",
//               "minimum": 1,
//               "default": 1
//             },
//             "release_date": {
//               "description": "The book's release date.",
//               "type": "string",
//               "format": "date"
//             },
//             "is_missing_implementation": {
//               "description": "The book is not (fully) implemented and thus needs to be excluded from the stable releases.",
//               "type": "boolean"
//             }
//           },
//           "required": ["name", "abbreviation"],
//           "additionalProperties": false
//         }
//       },
//       "minProperties": 1,
//       "additionalProperties": false
//     }
//   },
//   "required": ["id", "category", "age_restriction", "translations"],
//   "additionalProperties": false
// }

File: Sources/OptolithDatabaseSchema/GeneratedTypes/source/Publication.swift
//
//  Publication.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

public struct Publication: LocalizableEntity {
    /// The publication's identifier. An unique, increasing integer.
    public let id: Int
    
    /// The publication category.
    public let category: Category
    
    /// If the publication may contain adult content.
    public let containsAdultContent: Bool
    
    /// If the publication is not (fully) implemented and thus needs to be excluded from * stable releases.
    public let isMissingImplementation: Bool
    
    /// The specific other publications this publication depends on data from.
    public let prerequisites: PublicationPrerequisites?
    
    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<PublicationTranslation>

    public init(id: Int, category: Category, containsAdultContent: Bool, isMissingImplementation: Bool, prerequisites: PublicationPrerequisites? = nil, translations: LocaleMap<PublicationTranslation>) {
        self.id = id
        self.category = category
        self.containsAdultContent = containsAdultContent
        self.isMissingImplementation = isMissingImplementation
        self.prerequisites = prerequisites
        self.translations = translations
    }    
    
    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case category = "category"
        case containsAdultContent = "contains_adult_content"
        case isMissingImplementation = "is_missing_implementation"
        case prerequisites = "prerequisites"
        case translations = "translations"
    }
}

public struct PublicationTranslation: EntitySubtype {
    /// The publisher's publication identifier.
    public let id: String?
    
    /// The publication's name.
    public let name: NonEmptyString
    
    /// The publication's abbreviation.
    public let abbreviation: NonEmptyString
    
    /// The publication's release date.
    public let releaseDate: String?
    
    /// If this publication translation is not (fully) implemented and thus needs to be excluded from stable releases.
    public let isMissingImplementation: Bool

    public init(id: String? = nil, name: NonEmptyString, abbreviation: NonEmptyString, releaseDate: String? = nil, isMissingImplementation: Bool) {
        self.id = id
        self.name = name
        self.abbreviation = abbreviation
        self.releaseDate = releaseDate
        self.isMissingImplementation = isMissingImplementation
    }    
    
    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case name = "name"
        case abbreviation = "abbreviation"
        case releaseDate = "release_date"
        case isMissingImplementation = "is_missing_implementation"
    }
}

/// The publication category.
public enum Category: String, EntitySubtype {
    case coreRules = "CoreRules"
    case expansionRules = "ExpansionRules"
    case sourcebook = "Sourcebook"
    case regionalSourcebook = "RegionalSourcebook"
}
